capsules/src/radio.rs

```diff
@@ -188,35 +203,27 @@ impl<'a, R: radio::Radio> radio::TxClient for RadioDriver<'a, R> {
     }
 }
 
-#[allow(unused_variables)]
-impl<'a, R: radio::Radio> radio::RxClient for RadioDriver<'a, R> {
-    fn receive(&self, buf: &'static mut [u8], frame_len: u8, result: ReturnCode) {
+impl<'a, M: mac::Mac> mac::RxClient for RadioDriver<'a, M> {
+    fn receive<'b>(&self,
+                   buf: &'b [u8],
+                   /* We ignore the header because we pass the entire frame to
+                    * userspace */
+                   _: Header<'b>,
+                   data_offset: usize,
+                   data_len: usize,
+                   result: ReturnCode) {
         if self.app.is_some() {
-            self.app.map(move |app| {
-                if app.app_read.is_some() {
-                    let offset = self.radio.packet_payload_offset(buf) as usize;
-                    let len = self.radio.packet_get_length(buf) as usize;
-                    let dest = app.app_read.as_mut().unwrap();
-                    let d = &mut dest.as_mut();
-                    for (i, c) in buf[offset..offset + len].iter().enumerate() {
-                        d[i] = *c;
-                    }
-                    app.rx_callback
-                        .take()
-                        .map(|mut cb| { cb.schedule(usize::from(result), len, 0); });
-                }
-                self.radio.set_receive_buffer(buf);
+            self.app.map(move |app| if app.app_read.is_some() {
```

Comment:
If we did:

```rust
if let Some(dest) = app.app_read.as_mut() {
    ...
}
```

It would make me sleep much easier, even though I believe in practice they are the same.